src/Ext/Techno/Body.cpp

```diff
@@ -552,6 +553,337 @@ void TechnoExt::UpdateMindControlAnim(TechnoClass* pThis)
+	{
+		vPosH.X += RulesExt::Global()->Units_ShowHPOffset_WithoutShield.Get().X;
+		vPosH.Y += RulesExt::Global()->Units_ShowHPOffset_WithoutShield.Get().Y;
+	}
+
+	bool UseSHPShow = pTypeExt->HP_UseSHPShow.Get(RulesExt::Global()->Buildings_UseSHPShowHP);
+
+	if (UseSHPShow)
+	{
+		vPosH.X += 10;
+		vPosH.Y -= 27;
+		DrawBuildingSHPValue(pThis, pTypeExt, vPosH);
+	}
+	else
+	{
+		vPosH.X += 30;
+		DrawBuildingTextValue(pThis, pTypeExt, vPosH);
+	}
+}
+
+void TechnoExt::DrawBuildingTextValue(TechnoClass* pThis, TechnoTypeExt::ExtData* pTypeExt, Point2D vPos)
+{
+	wchar_t Healthpoint[0x20];
+	swprintf_s(Healthpoint, L"%d/%d", pThis->Health, pThis->GetTechnoType()->Strength);
+
+	COLORREF ShowHPColor;
+
+	if (pThis->IsGreenHP())
+	{
+		Vector3D<int>ShowHPColorHighVector3D = pTypeExt->HP_ShowColorHigh.Get(RulesExt::Global()->Buildings_ShowColorHPHigh.Get());
+		COLORREF ShowHPColorHigh = Drawing::RGB2DWORD(ShowHPColorHighVector3D.X, ShowHPColorHighVector3D.Y, ShowHPColorHighVector3D.Z);
+		ShowHPColor = ShowHPColorHigh;
+	}
+	else if (pThis->IsYellowHP())
+	{
+		Vector3D<int>ShowHPColorMidVector3D = pTypeExt->HP_ShowColorMid.Get(RulesExt::Global()->Buildings_ShowColorHPMid.Get());
+		COLORREF ShowHPColorMid = Drawing::RGB2DWORD(ShowHPColorMidVector3D.X, ShowHPColorMidVector3D.Y, ShowHPColorMidVector3D.Z);
+		ShowHPColor = ShowHPColorMid;
+	}
+	else
+	{
+		Vector3D<int>ShowHPColorLowVector3D = pTypeExt->HP_ShowColorLow.Get(RulesExt::Global()->Buildings_ShowColorHPLow.Get());
+		COLORREF ShowHPColorLow = Drawing::RGB2DWORD(ShowHPColorLowVector3D.X, ShowHPColorLowVector3D.Y, ShowHPColorLowVector3D.Z);
+		ShowHPColor = ShowHPColorLow;
+	}
+
+	bool ShowBackground = pTypeExt->HP_ShowBackground.Get(RulesExt::Global()->Buildings_ShowBackground.Get());
+
+	RectangleStruct rect = { 0,0,0,0 };
+	DSurface::Temp->GetRect(&rect);
+	COLORREF BackColor = 0;
+	TextPrintType PrintType = TextPrintType(int(TextPrintType::Right) + (ShowBackground ? int(TextPrintType::GradAll) : 0));
+
+	//DSurface::Temp->DrawText(Healthpoint, vPosH.X, vPosH.Y, ShowHPColor);
+	DSurface::Temp->DrawTextA(Healthpoint, &rect, &vPos, ShowHPColor, BackColor, PrintType);
+}
+
+void TechnoExt::DrawBuildingSHPValue(TechnoClass* pThis, TechnoTypeExt::ExtData* pTypeExt, Point2D vPos)
+{
+	const int Strength = pThis->GetTechnoType()->Strength;
+	const int Health = pThis->Health;
+
+	const DynamicVectorClass<char>StrengthVector = IntToVector(Strength);
+	const DynamicVectorClass<char>HealthVector = IntToVector(Health);
+
+	const int Length = StrengthVector.Count + HealthVector.Count + 1;
+
+	const int Interval = pTypeExt->HP_ShowInterval.Get(RulesExt::Global()->Buildings_HPNumberInterval);
+
+	SHPStruct* NumberSHP = pTypeExt->SHP_HealthSHP;
+	if (NumberSHP == nullptr)
+	{
+		char FilenameSHP[0x20];
+		strcpy_s(FilenameSHP, pTypeExt->HP_ShowSHP.data());
+		if (strcmp(FilenameSHP, "") == 0)
+			strcpy_s(FilenameSHP, RulesExt::Global()->Buildings_HPNumberSHP.data());
```

Comment:
Nullable<T>::Get(T default) could simplify this code.
```suggestion
		char FilenameSHP[0x20];
		strcpy_s(FilenameSHP, pTypeExt->HP_ShowSHP.Get(RulesExt::Global()->Buildings_HPNumberSHP).data());
```